cmd/search: Add --job-uri-prefix

wking opened the pull request:
... and pass it around, instead of hard-coding the prefix in a number of locations.  As part of consuming the prefix URI, extend `MetadataFor` to handle the path decoding and job-URI construction as well, which
allows us to centralize some code that was previously duplicated between `renderWithContext` and `renderSummary`.

As part of this change, I've adjusted the output format a bit.  Most obviously by removing `template.HTMLEscapeString(string(parts[3])` from some `renderWithContext output`.  For example, master produced entries like:

```
build-log.txt from PR 22369 pull-ci-openshift-origin-master-e2e-aws-serial #5405 About an hour
build-log.txt from PR openshift_installer pull-ci-openshift-installer-master-e2e-openstack #189 4 hours
build-log.txt from build release-openshift-origin-installer-e2e-aws-upgrade #859 3 hours
```

With this commit, we get entries like:

```
build-log from pull pull-ci-openshift-origin-master-e2e-aws #7812 {duration}
build-log from pull pull-ci-openshift-installer-master-e2e-aws #5542 {duration}
build-log from build release-openshift-origin-installer-e2e-aws-4.0 #7077 {duration}
```

The PR number is nice, but when the PR number is instead a repository slug, it's redundant with the job name.  In fact, the trigger string is also fairly redundant with the job name (e.g. `...from pull pull-ci-...`, but I'm leaving that alone for now because at least we consistently detect it (vs. flip-flopping between PR number and repo slugs).

I don't have durations above, because when I copied down some example data from the `ci-search` pod, I didn't get whatever holds the finished-at times.  Or I broke duration calculation somehow with this commit ;).

wking commented:
/assign @smarterclayton

smarterclayton commented:
You don't need perms to run the exact indexer - it runs unauthenticated, so you should be able to index locally and verify duration is correct.

wking commented:
#22 fixes the durations, so I've rebased this on top, renamed the option to the provider-agnostic `--job-uri-prefix`, and renamed `--deck-url` to `--deck-uri` for consistency (we don't care if these are generically locate-able or not) with 16c56d9 -> b70d805.  Now durations work.

wking commented:
Rebased around #20 with b70d805 -> 098d2cc.

smarterclayton commented:
Did you also open a PR to release that will change it there so we don't break search when this merges?

wking commented:
> Did you also open a PR to release that will change it there so we don't break search when this merges?

I've added 0a61731 here so we can drop the `--deck-url` option in the release repo, and put up openshift/release#4066 (with a `/hold`) to make that drop.

wking commented:
Cherry-picked #21 in with 41e4e85.

wking commented:
Landed via #24.
